react/Popup/index.jsx

```diff
@@ -0,0 +1,199 @@
+
+  var width = window.innerWidth
+    ? window.innerWidth
+    : document.documentElement.clientWidth
+    ? document.documentElement.clientWidth
+    : screen.width
+  var height = window.innerHeight
+    ? window.innerHeight
+    : document.documentElement.clientHeight
+    ? document.documentElement.clientHeight
+    : screen.height
+
+  var left = width / 2 - w / 2 + dualScreenLeft
+  var top = height / 2 - h / 2 + dualScreenTop
+  return {
+    w,
+    h,
+    top,
+    left
+  }
+}
+
+/**
+ * Renders a popup and listen to popup events
+ */
+export class Popup extends PureComponent {
+  constructor(props, context) {
+    super(props, context)
+
+    // url is not expected to change
+    const { initialUrl } = props
+    this.state = {
+      url: initialUrl
+    }
+
+    this.handleClose = this.handleClose.bind(this)
+    this.handleMessage = this.handleMessage.bind(this)
+    this.handleLoadStart = this.handleLoadStart.bind(this)
+  }
+
+  componentDidMount() {
+    this.showPopup()
+  }
+
+  componentWillUnmount() {
+    this.killPopup()
+  }
+
+  addListeners(popup) {
+    // Listen here for message FROM popup
+    window.addEventListener('message', this.handleMessage)
+
+    // rest of instructions only on mobile app
+    if (!isMobileApp()) return
+    popup.addEventListener('loadstart', this.handleLoadStart)
+    popup.addEventListener('exit', this.handleClose)
+  }
+
+  removeListeners() {
+    const { popup } = this.state
+    window.removeEventListener('message', this.handleMessage)
+
+    // rest of instructions only if popup is still opened
+    if (popup.closed) return
+
+    // rest of instructions only on mobile app
+    if (!isMobileApp()) return
+    popup.removeEventListener('loadstart', this.handleLoadStart)
+    popup.removeEventListener('exit', this.handleClose)
+  }
+
+  handleMessage(messageEvent) {
+    const { popup } = this.state
+    const { onMessage } = this.props
+    const isFromPopup = popup === messageEvent.source
+    if (isFromPopup && typeof onMessage === 'function') onMessage(messageEvent)
+  }
+
+  handleClose() {
+    const { popup } = this.state
+
+    this.killPopup()
+
+    const { onClose } = this.props
+    if (typeof onClose === 'function') onClose(popup)
+  }
+
+  showPopup() {
+    const { height, width, title } = this.props
+    const { url } = this.state
+    const { w, h, top, left } = popupCenter(width, height)
+    /**
+     * ATM we also use window.open on Native App in order to open
+     * InAppBrowser. But some provider (Google for instance) will
+     * block us. We need to use a SafariViewController or Chrome Custom Tab.
+     * So
+     */
+    const popup = window.open(
+      url,
+      title,
+      `scrollbars=yes, width=${w}, height=${h}, top=${top}, left=${left}`
+    )
+    // Puts focus on the newWindow
+    if (popup.focus) {
+      popup.focus()
+    }
+
+    this.addListeners(popup)
+    this.startMonitoringClosing(popup)
+    this.setState({ popup })
```

Comment:
Can we store the popup in `this.popup` instead of in the state? Extra re-renders in not a big concern since `render` returns `null`, but if we can avoid storing things unrelated to rendering in state, it would be nice.